src/ast/nodes/MemberExpression.ts

```diff
@@ -2,53 +2,55 @@ import relativeId from '../../utils/relativeId';
 import { ExpressionNode, Node, NodeBase } from './shared/Node';
 import Variable from '../variables/Variable';
 import ExecutionPathOptions from '../ExecutionPathOptions';
-import { isLiteral } from './Literal';
 import CallOptions from '../CallOptions';
 import MagicString from 'magic-string';
-import Identifier, { isIdentifier } from './Identifier';
+import Identifier from './Identifier';
 import NamespaceVariable from '../variables/NamespaceVariable';
 import ExternalVariable from '../variables/ExternalVariable';
 import { ForEachReturnExpressionCallback, SomeReturnExpressionCallback } from './shared/Expression';
 import { NodeType } from './NodeType';
 import { NodeRenderOptions, RenderOptions } from '../../utils/renderHelpers';
-import { isUnknownKey, ObjectPath, ObjectPathKey, UNKNOWN_KEY } from '../values';
+import {
+	LiteralValueOrUnknown,
+	ObjectPath,
+	ObjectPathKey,
+	UNKNOWN_KEY,
+	UNKNOWN_VALUE
+} from '../values';
 import { BLANK } from '../../utils/blank';
+import Literal from './Literal';
 
-const validProp = /^[a-zA-Z_$][a-zA-Z_$0-9]*$/;
-
-function getPropertyKey(memberExpression: MemberExpression): ObjectPathKey {
+function getPropertyKey(memberExpression: MemberExpression): string | null {
 	return memberExpression.computed
 		? getComputedPropertyKey(memberExpression.property)
 		: (<Identifier>memberExpression.property).name;
 }
 
-function getComputedPropertyKey(propertyKey: ExpressionNode): ObjectPathKey {
-	if (isLiteral(propertyKey)) {
-		const key = String(propertyKey.value);
-		return validProp.test(key) ? key : UNKNOWN_KEY;
+function getComputedPropertyKey(propertyKey: ExpressionNode): string | null {
+	if (propertyKey instanceof Literal) {
+		return String(propertyKey.value);
 	}
-	return UNKNOWN_KEY;
+	return null;
 }
 
 type PathWithPositions = { key: string; pos: number }[];
 
 function getPathIfNotComputed(memberExpression: MemberExpression): PathWithPositions | null {
 	const nextPathKey = memberExpression.propertyKey;
 	const object = memberExpression.object;
-	if (isUnknownKey(nextPathKey)) {
-		return null;
-	}
-	if (isIdentifier(object)) {
-		return [
-			{ key: object.name, pos: object.start },
-			{ key: nextPathKey, pos: memberExpression.property.start }
-		];
-	}
-	if (isMemberExpression(object)) {
-		const parentPath = getPathIfNotComputed(object);
-		return (
-			parentPath && [...parentPath, { key: nextPathKey, pos: memberExpression.property.start }]
-		);
+	if (typeof nextPathKey === 'string') {
+		if (object instanceof Identifier) {
+			return [
+				{ key: object.name, pos: object.start },
+				{ key: nextPathKey, pos: memberExpression.property.start }
+			];
+		}
+		if (isMemberExpression(object)) {
+			const parentPath = getPathIfNotComputed(object);
+			return (
+				parentPath && [...parentPath, { key: nextPathKey, pos: memberExpression.property.start }]
+			);
+		}
```

Comment:
This is only called during initialization and my feeling is that due to the relation to the namespace logic, this should not be able to handle computed properties except in a few special cases due to the dynamic nature of the typing system. But this could definitely be improved and unified in the future!